packages/core/src/connectivity/scp-cf/destination/destination-from-service.ts

```diff
@@ -220,21 +226,69 @@ class DestinationFromServiceRetriever {
     return credentials[0];
   }
 
-  private async addOAuthSamlAuth(
-    destination: Destination,
-    destinationOrigin: DestinationOrigin
+  private fetchDestinationByToken(
+    jwt: string | AuthAndExchangeTokens
   ): Promise<Destination> {
-    const destinationService = getService('destination');
+    const destinationService = getDestinationService();
 
-    if (!destinationService) {
+    return fetchDestination(
+      destinationService.credentials.uri,
+      jwt,
+      this.name,
+      this.options
+    );
+  }
+
+  private async addOAuth2UserTokenExchange(
+    destinationOrigin: DestinationOrigin
+  ): Promise<Destination> {
+    if (!this.options.userJwt) {
       throw Error(
-        `Failed to fetch destination "${this.name}"! No binding to a destination service found.`
+        'No user token (JWT) has been provided! This is strictly necessary for OAuth2UserTokenExchange.'
+      );
+    }
+    // This covers the three OAuth2UserTokenExchange cases https://help.sap.com/viewer/cca91383641e40ffbe03bdc78f00f681/Cloud/en-US/39d42654093e4f8db20398a06f7eab2b.html
+
+    // Case 1 Destination in provider and jwt issued for provider account
+    if (this.isProviderAndSubscriberSameTenant()) {
+      const authHeaderJwt = await userApprovedServiceToken(
```

Comment:
This file is already pretty huge. I think this would be a good point to consider factoring out some functions. Maybe some methods could be turned into pure functions and moved out of this class?

Reply:
I agree. However, I tried this for a bit and somehow the methods are always related and it would be a bigger effort. So I would do this in a seperate PR.